File: database/model/node.go
package model

const (
	ProxyModeNone     = "none"
	ProxyModeProxyUrl = "proxyUrl"
	ProxyModeOutbound = "outbound"
)

const (
	NodeStatusUnknown = "unknown"
	NodeStatusOnline  = "online"
	NodeStatusOffline = "offline"
)

type Node struct {
	Id       int    `json:"id" form:"id" gorm:"primaryKey;autoIncrement"`
	Remark   string `json:"remark" form:"remark"`
	Address  string `json:"address" form:"address"`
	BasePath string `json:"basePath" form:"basePath"`
	Username string `json:"username" form:"username"`
	Password string `json:"password" form:"password"`

	ProxyMode   string `json:"proxyMode" form:"proxyMode"`
	ProxyUrl    string `json:"proxyUrl" form:"proxyUrl"`
	OutboundTag string `json:"outboundTag" form:"outboundTag"`

	Enable bool `json:"enable" form:"enable"`

	// Port is the listen port of this node's own default inbound, built from the
	// shared universal template when the node is added (and its connection is OK).
	Port int `json:"port" form:"port"`
	// RemoteInboundId is the id of the default inbound created on this node
	// (0 = not created yet). Client add/edit operations target this inbound.
	RemoteInboundId int `json:"remoteInboundId" form:"remoteInboundId"`
	// InboundError holds the last error from creating this node's default inbound.
	InboundError string `json:"inboundError" form:"inboundError"`
	// LastSync is the last successful traffic-sync time for this node (unix ms).
	LastSync int64 `json:"lastSync" form:"lastSync"`

	Status    string `json:"status" form:"status"`
	LastCheck int64  `json:"lastCheck" form:"lastCheck"`
	LastError string `json:"lastError" form:"lastError"`
}

// NodeSharedInbound is the template used to create the same inbound on every node.
type NodeSharedInbound struct {
	Id             int    `json:"id" form:"id" gorm:"primaryKey;autoIncrement"`
	Remark         string `json:"remark" form:"remark"`
	Protocol       string `json:"protocol" form:"protocol"`
	Port           int    `json:"port" form:"port"`
	Settings       string `json:"settings" form:"settings"`
	StreamSettings string `json:"streamSettings" form:"streamSettings"`
	Sniffing       string `json:"sniffing" form:"sniffing"`
}

// NodeSharedInboundMap records which remote inbound id corresponds to the
// shared inbound on a given node.
type NodeSharedInboundMap struct {
	Id              int `json:"id" form:"id" gorm:"primaryKey;autoIncrement"`
	SharedInboundId int `json:"sharedInboundId" form:"sharedInboundId"`
	NodeId          int `json:"nodeId" form:"nodeId"`
	RemoteInboundId int `json:"remoteInboundId" form:"remoteInboundId"`
}

// NodeSharedClient is the single client tracked on the shared inbound across all nodes.
type NodeSharedClient struct {
	Id              int    `json:"id" form:"id" gorm:"primaryKey;autoIncrement"`
	SharedInboundId int    `json:"sharedInboundId" form:"sharedInboundId"`
	ClientId        string `json:"clientId" form:"clientId"`
	Email           string `json:"email" form:"email"`
	TotalGB         int64  `json:"totalGB" form:"totalGB"`
	ExpiryTime      int64  `json:"expiryTime" form:"expiryTime"`
	Enable          bool   `json:"enable" form:"enable"`
}

// NodeClientTrafficSnapshot stores the last-seen traffic counters for a
// (node, email) pair, used to compute usage deltas between sync ticks.
type NodeClientTrafficSnapshot struct {
	Id     int    `json:"id" gorm:"primaryKey;autoIncrement"`
	NodeId int    `json:"nodeId" gorm:"uniqueIndex:idx_node_email"`
	Email  string `json:"email" gorm:"uniqueIndex:idx_node_email"`
	Up     int64  `json:"up"`
	Down   int64  `json:"down"`
}

File: web/assets/js/model/node.js
const ProxyModes = {
    NONE: 'none',
    PROXY_URL: 'proxyUrl',
    OUTBOUND: 'outbound',
};

const NodeStatus = {
    UNKNOWN: 'unknown',
    ONLINE: 'online',
    OFFLINE: 'offline',
};

class XNode {
    constructor(data) {
        this.id = 0;
        this.remark = '';
        this.address = '';
        this.basePath = '';
        this.username = '';
        this.password = '';
        this.proxyMode = ProxyModes.NONE;
        this.proxyUrl = '';
        this.outboundTag = '';
        this.enable = true;
        this.port = RandomUtil.randomIntRange(10000, 60000);
        this.remoteInboundId = 0;
        this.inboundError = '';
        this.lastSync = 0;
        this.status = NodeStatus.UNKNOWN;
        this.lastCheck = 0;
        this.lastError = '';

        if (data == null) {
            return;
        }
        ObjectUtil.cloneProps(this, data);
    }
}

class SharedClient {
    constructor(data) {
        this.id = RandomUtil.randomUUID();
        this.email = RandomUtil.randomLowerAndNum(8);
        this.totalGB = 0;
        this.expiryTime = 0;
        this.enable = true;

        if (data == null) {
            return;
        }
        if (data.clientId) this.id = data.clientId;
        if (data.id) this.id = data.id;
        if (data.email != null) this.email = data.email;
        if (data.totalGB != null) this.totalGB = data.totalGB;
        if (data.expiryTime != null) this.expiryTime = data.expiryTime;
        if (data.enable != null) this.enable = data.enable;
    }

    get _totalGB() {
        return toFixed(this.totalGB / ONE_GB, 2);
    }

    set _totalGB(gb) {
        this.totalGB = toFixed(gb * ONE_GB, 0);
    }

    get _expiryTime() {
        if (this.expiryTime === 0 || this.expiryTime === '') {
            return null;
        }
        return moment(this.expiryTime);
    }

    set _expiryTime(t) {
        this.expiryTime = (t == null || t === '') ? 0 : t.valueOf();
    }

    toClient() {
        return {
            id: this.id,
            password: this.id,
            flow: '',
            email: this.email,
            totalGB: this.totalGB,
            expiryTime: this.expiryTime,
            enable: this.enable,
            tgId: '',
            subId: RandomUtil.randomLowerAndNum(16),
            reset: 0,
        };
    }
}
